imports/client/components/SplitPanePlus.tsx

```diff
@@ -111,168 +154,175 @@ class SplitPanePlus extends React.Component<SplitPanePlusProps, SplitPanePlusSta
     });
-    this.attemptCollapse(size);
-    if (this.props.collapsed === 0) {
-      this.recordSize(size);
-    }
-    if (this.props.onDragFinished) {
-      this.props.onDragFinished(size, this.props.collapsed);
+    // May be called with a number or a string representing a percentage
+    let size: number = Number(rawSize);
+    if (typeof rawSize === 'string') {
+      const rawSizeAsPercent: number = Number(rawSize.slice(0, -1));
+      if (rawSize.slice(-1) === '%' && !Number.isNaN(rawSizeAsPercent)) {
+        size = (rawSizeAsPercent * this.measure(this.splitPaneNode())) / 100.0;
+      }
     }
-  }
-
-  recordSize(size: number) {
-    this.setState({
-      lastSize: size,
-      lastRelSize: size / this.measure(this.splitPaneNode()!),
-    });
-  }
-
-  attemptCollapse(size: number) {
-    const oldCollapsed = this.props.collapsed;
-    const newCollapsed = this.calculateCollapse(size);
-    if (oldCollapsed !== newCollapsed && this.props.onCollapseChanged) {
-      this.props.onCollapseChanged(newCollapsed);
+    if (!Number.isNaN(size) && this.props.onPaneChanged) {
+      this.props.onPaneChanged(size, this.calculateCollapse(size), 'drag');
     }
   }
 
   calculateCollapse(size: number) {
-    const fullSize = this.measure(this.splitPaneNode()!);
-    const halfResizerSize = this.measure(this.resizerNode()!) / 2;
-    let autoCollapsePrimary = this.props.autoCollapse1!;
-    let autoCollapseSecondary = this.props.autoCollapse2!;
+    const fullSize = this.measure(this.splitPaneNode());
+    let autoCollapsePrimary = this.props.autoCollapse1;
+    let autoCollapseSecondary = this.props.autoCollapse2;
     if (this.props.primary !== 'first') {
-      autoCollapsePrimary = this.props.autoCollapse2!;
-      autoCollapseSecondary = this.props.autoCollapse1!;
+      autoCollapsePrimary = this.props.autoCollapse2;
+      autoCollapseSecondary = this.props.autoCollapse1;
     }
-    if (size + halfResizerSize <= autoCollapsePrimary) {
-      // Collapse Primary
+    if (autoCollapsePrimary >= 0 && size <= autoCollapsePrimary) {
       return this.props.primary === 'first' ? 1 : 2;
-    } else if (size + halfResizerSize >= fullSize - autoCollapseSecondary) {
-      // Collapse Secondary
+    } else if (autoCollapseSecondary >= 0 && size >= fullSize - autoCollapseSecondary) {
       return this.props.primary === 'first' ? 2 : 1;
     }
     return 0;
   }
 
   splitPaneNode(): HTMLElement | null {
-    if (!this.ref.current) {
-      return null;
-    }
-    if (!(this.ref.current.firstChild instanceof HTMLElement)) {
+    if (!this.ref.current || !(this.ref.current.firstChild instanceof HTMLElement)) {
       return null;
     }
     return this.ref.current.firstChild;
   }
 
-  primaryPaneNode() {
+  primaryPaneNode(): HTMLElement | null {
     const root = this.splitPaneNode();
     const className = `Pane${this.props.primary === 'first' ? 1 : 2}`;
     return root ? _.find(root.childNodes, (n) => (
       n instanceof Element &&
```

Comment:
I think these may also need to change to `HTMLElement` for the cast below to be safe

Reply:
Not entirely sure what you mean by this. Maybe you misread the ternary? If so, I can't blame you because this is terrible. Cleaned it up dramatically.